src/iface/fragments.rs

```diff
@@ -0,0 +1,170 @@
+            id: 0,
+            last_used: Instant::from_millis(0),
+            header: false,
+            header_len: 0,
+            total_len: 0,
+        }
+    }
+
+    /// Reset packet
+    pub fn reset(&mut self) {
+        self.id = 0;
+        self.last_used = Instant::from_millis(0);
+        self.header = false;
+        self.header_len = 0;
+        self.total_len = 0;
+        self.assembler.remove_front();
+    }
+
+    /// Add fragment into the packet
+    pub fn add(&mut self, header_len: usize, offset: usize, payload_len: usize, data: &[T], time: Instant) -> Result<(), ()>
+        where T: Clone
+    {
+        debug_assert!(time >= self.last_used);
+        self.last_used = time;
+
+        if !self.header { 
+            self.assembler.add(0, header_len)?;
+            let range = 0..header_len;
+            self.rx_buffer[range].clone_from_slice(&data[0..header_len]);
+            self.header_len = header_len;
+            self.header = true;
+        }
+
+        self.assembler.add(offset+self.header_len, payload_len)?;
+        let range = offset+self.header_len..offset+payload_len+self.header_len;
+        self.rx_buffer[range].clone_from_slice(&data[header_len..]);
+        Ok(())
+    }
+
+    /// Return `true` if the packet has all fragments,
+    /// and can be reassmbled, and `false` otherwise
+    pub fn check_contig_range(&self) -> bool {
+        if self.total_len != 0 {
+            match self.assembler.peek_front() {
+                Some(front) => {
+                    if front == self.total_len {
+                        return true;
+                    }
+                }
+                None => {       
+                }
+            }
+        }
+        false
+    }
+
+    pub fn front(&mut self) -> Option<usize> {
+        self.assembler.remove_front()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.assembler.is_empty()
+    }
+
+    pub fn set_id(&mut self, id: u16) {
+        self.id = id;
+    }
+
+    pub fn set_total_len(&mut self, len: usize) {
+        self.total_len = len;
+    }
+}
+
+pub struct Set<'a> {
+	packets: ManagedSlice<'a, Packet<'a, u8>>,    
+}
+
+impl<'a> Set<'a> {
+	/// Create a new set of packets
+    pub fn new<S>(storage: S) -> Set<'a>
+	    where S: Into<ManagedSlice<'a, Packet<'a, u8>>>,
+    {
+        Set { packets: storage.into() }
+    }
+    
+    /// Add new packet into the set
+    pub fn add(&mut self, new_packet: Packet<'a, u8>) {
+    	match self.packets {
+    		ManagedSlice::Borrowed(ref mut packets) => {
+		    	for packet in packets.iter_mut() {
+		            if packet.id == 0 {
+		            	// empty packet
+		            	*packet = new_packet;
+		            	return;
+		            }
+		        }		
+    		},
+    		#[cfg(any(feature = "std", feature = "alloc"))]
+    		ManagedSlice::Owned(ref mut packets) => {
+    			packets.push(new_packet);
+    		}
+    	}
+    }
+
+    /// Check if the set contains a packet with given `id`
```

Comment:
This isn't enough. You need to at least match the ident field with the source IP, because otherwise you're introducing a vulnerability: anyone who knows that there is a flow of fragmented packets has an 1/65536 chance of arbitrarily altering their contents. This is a 100% viable attack vector and very similar attacks (utilizing the port number, but the concept is the same) have been demonstrated against DNS two decades ago.

Reply:
Also, you probably should use a ManagedMap here instead of hardcoding a linear search.

Reply:
Now I am checking the id and src & dst addresses. 
ManagedMap is not possible here because each packet needs an underlying storage. So I cannot simply add a new packet to the map, unless I explicitly provide a storage for that packet.

Reply:
Of course it is possible, you use a map between ident/addresses and the *index* into the packet storage.